### PR TITLE
Add ids to session subjects

### LINKED ISSUE #253: Add unique subject id
When creating a new session in the DB, add an `_id` field of type `ObjectId` to the `subject` subdocument.

If another session with the same `subject.code` already exists in the new session's project, copy the `subject._id` from the existing session, else generate a new one.

Also, allow `subject._id` to be set explicitly on new and existing sessions via `/sessions`.

### PR DESCRIPTION
**Pull request by nagem**
Closes #253, check out the issue for information about the requirements.

### CODE REVIEW

File: api/dao/containerutil.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import bson.objectid
```
Comment (nagem) [inline api/dao/containerutil.py]: This one broke the camel's back of needing a better place to put functions that act on specific containers and interact with the db. Suggesting this module as a midstep between ad hoc as needed methods outside of the `doa` package and having container-specific modules.

File: api/dao/containerutil.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+    Use the same _id as other subjects in the session's project with the same code
+    If no _id is found, generate a new _id
+    """
+    result = None
+    if subject is None:
+        subject = {}
+    if subject.get('_id') is not None:
+        return subject
+    if subject.get('code') is not None:
```
Comment (kofalt) [inline api/dao/containerutil.py]: Suggest adding whitespace before this line to separate out the conditionals for readability

File: api/dao/containerutil.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        query = {'subject.code': subject['code'],
+                 'project': pid,
```
Comment (kofalt) [inline api/dao/containerutil.py]: `subject` is checked for `None`, while `pid` is not. Intended?

**Comment by kofalt** [comment]
LGTMs after comment

File: api/dao/containerutil.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+                 'subject._id': {'$exists': True}}
+        result = config.db.sessions.find_one(query)
+    if result is not None:
+        subject['_id'] = str(result['subject']['_id'])
+    else:
+        subject['_id'] = str(bson.ObjectId())
+    return subject
```
Comment (gsfr) [inline api/dao/containerutil.py]: Missing newline at EOF.

Also surprised to see that you are storing the `ObjectId` as a string in the DB. Why?

Comment (nagem) [inline api/dao/containerutil.py]: Good catch, I was missing `"ensure_newline_at_eof_on_save": true` in my Sublime prefs.

I was playing with the idea of keeping the `_id` field as a string to cut back on having to ensure it is a bson ObjectId on every update/create of a session object throughout the codebase. After thinking about it, it's now back to being a bson ObjectId.

**Comment by nagem** [comment]
Rebased and retested, will merge tomorrow unless there are any new comments.

**Pull request merged**